### ISSUE TITLE
blt deploy should not ignore composer platform requirements

### ISSUE THREAD

**Issue by christopher-hopper** [acquia/blt#3417]
My system information:
* Operating system type: Linux
* Operating system version: Ubuntu 16.04 LTS
* BLT version: 9.2.x

When I run this command:
```
blt deploy
```

On a desktop computer running PHP 5.6, 
I got a package installed that was incompatible with PHP 7.2

### Related to

#3304 
#3305

**Comment by christopher-hopper** [comment]
BLT should not ignore platform requirements when building a production ready build artefact. 

If an individual developer has a need to run multiple versions of PHP, or use an incompatible PHP version during a build, there are resources available to support that person. Here are some suggestions:

1. Use phpenv to switch between multiple PHP versions on a CI server https://github.com/phpenv/phpenv-installer  or locally on macOS
2. Use the [composer platform config](https://getcomposer.org/doc/06-config.md#platform) to fake platform packages (PHP and extensions) so that you can emulate a production env or define your target platform in the config. 

I recommend we revert the change to ignore platform requirements during the deploy build process.

**Comment by christopher-hopper** [comment]
If you want to add an additional check, into the Deploy command, you could call the following to identify incompatibilities before running the composer install.

https://getcomposer.org/doc/03-cli.md#check-platform-reqs

```
# composer check-platform-reqs
composer-plugin-api  1.1.0     success
php                  7.1.26    success
```

**Comment by danepowell** [comment]
I don't understand how ignoring platform requirements would negatively impact you. When you run `blt deploy`, it runs `composer install`, which installs the exact versions of packages defined by `composer.lock`. Regardless of whether you use `--ignore-platform-reqs` or not, you should get the exact same set of packages installed.

The only way I could see `--ignore-platform-reqs` changing what gets installed is if you run `composer update` or don't have an existing `composer.lock` file at all.

While I agree that in principle, people should run the same versions of PHP locally and remotely, as a practical matter I just don't understand how this would cause problems. Let me know what I'm missing.

**Comment by danepowell** [comment]
Closing due to lack of response.

**Comment by christopher-hopper** [comment]
@danepowell 
Yes, the composer.lock file SHOULD lock packages to the installable set that worked for the last person who ran an install or update. Where we run into problems is when we have inexperienced developers or non-colocated teams with varying skill levels who are working to tight deadlines: this is the BLT audience. 

People do some strange things, just to get projects over the line. Things you and I would gasp at. I've had an agency team once just delete the lock file entirely, so they could get their code to deploy. 

I read the issue that lead to this `--ignore-platform-reqs` being added.

#3304 :

> I am running BLT on my local machine which has PHP 7.3 (has no mcrypt) and deploying to an environment with PHP 7.1 (which has mcrypt). 

With the forced switchover between PHP 7.1 -> 7.2 upon us soon we could be supporting people to understand the platform requirements and use them correctly, rather than putting in exceptions so someone can install and deploy code using PHP 7.3 for a platform running PHP 7.1.

**Comment by christopher-hopper** [comment]
Here's a scenario:

1. Bob, the architect adds a platform requirement to the project to ensure all packages are php 7.2 compatible.
2. Nick, joins the team, and he is using php 7.3 because that's what homebrew gave him
3. Nick tries to install a package but it tells him he can't because of platform requirements. 
4. Nick looks on stackexchange and learns he can force the install by using `composer require --ignore-platform-reqs`, which works.
5. The package is now installed with a new language feature that is available in PHP 7.3.
6. Nick goes to deploy his code using BLT. Because BLT also ignore platform requirements, nothing stops him from deploy his code to a platform that runs PHP 7.2.
7. Bob gets notified that the Dev server has gone down by one of the team.

**Comment by danepowell** [comment]
Thanks for following up. I'm going to reopen this to get more input, as I'm on the fence about what to do. I just saw your response to @hussainweb over here, since I wasn't originally on that thread: https://github.com/acquia/blt/issues/3304#issuecomment-469133350

> Nick looks on stackexchange and learns he can force the install

It doesn't make sense to enforce this _just_ to stop people from shooting themselves in the foot by blindly copying code from stackexchange. If you are following any sane workflow (I understand not everyone does), this should never be a problem. Having said that, if we _can_ help people to avoid this in service of a better overall architecture, then obviously that's a good thing.

> The solution to this issue is for @hussainweb to utilize a tool like phpenv to manage his local PHP versions correctly

I _mostly_ agree with this, with some caveats. Sure it makes sense that folks should run the same version of PHP locally and remotely, but in practice this _usually_ isn't a problem (although sometimes it is, as with mcrypt and the syntax changes from 5.6 to 7.) And if you are on a bunch of different projects at once, constantly switching PHP versions is a pain.

On the flip side, one could make the argument that when you are deploying, you are not _running_ your site, you are _building_ it, so there's no need to have exactly the same version of runtime dependencies (e.g. PHP) installed as long as you get the same build artifact. We might break people's CI tools by changing this, and it sets an overall precedent of runtime and buildtime environments being identical. I _assume_ most people build via the same environment they test in, making this little more than a pedantic distinction, but I could be wrong.

If we make this change, I wonder if we should also do more to encourage people to use PHP versions consistently (via composer, BLT validation, documentation on how to use phpenv, or something else).

**Comment by hussainweb** [comment]
Thank you for this discussion. I read through the post and the scenario and still think that ignoring platform requirements should be supported. I don't mind if it is not the default behaviour. If `blt deploy` can take a parameter to ignore platform requirements which it can pass on to `composer`, that works for me. Of course, we have to be careful as this will break existing CI scripts.

Having said that, I think that the scenarios in which `blt deploy` is used always involve a lock file (or at least it should). The issues I see in the OP and scenario happen only when installing without a lock file, or when the lock file is "corrupted" by an incorrect `composer require` command. I know that we should make it easier on developers but the action where this happens is out of BLT's scope, IMHO. I know removing `--ignore-platform-reqs` would work but that seems to me to be fixing the symptom, not the problem. Again, that may not be such a bad thing.

I would like to see a different solution to this problem (maybe something like checking if `composer.lock` exists before deploying). At the same time, I think ignoring platform requirements is something that should be supported, even if not by default. I think it does make sense that the build system would not have the extensions that are needed to run the site and that shouldn't stop installing the site.

**Comment by christopher-hopper** [comment]
I know this all sounds fairly theoretical, but the driver for me was past personal experience when we moved from PHP 5.6 to PHP 7.1. The following helped me enforce sanity for my team at the time after we had PHP 5.6 code sneak back into a build artefact:

```
composer require php:~7.1.0
```

Let's be okay with people deciding to consciously ignore the warnings, if that's what they want to do, but it'd be better not to circumvent the warnings to appease edge case users. 

Matching platform requirements is an important part of how composer works, in addition to the semantic version matching. [This post sums up the key concepts](https://andy-carter.com/blog/composer-php-platform) really well I think.

[Enforcing a PHP Version for Installed Composer Packages](https://andy-carter.com/blog/composer-php-platform)

**Comment by danepowell** [comment]
Automated testing should have caught both of the scenarios you've outlined above (rogue developers using --ignore-platform-reqs or deleting composer.lock). If you are deploying by hand without enforcing automated testing, then I don't think the deploy step is the right place to start doing even more validation and testing. And even if we did, people might Google your post above about how to fake platform requirements and shoot themselves in the foot anyway :smile: 

Having said that: I think I see a happy medium here, which would be to add fairly loose PHP version requirements to the template composer.json (i.e. `^7.1`, same as BLT's), and remove `--ignore-platform-reqs` from the deploy step by default, _but_ make it an optional parameter if people really know what they are getting into.

This would encourage people to at least consider PHP versions, without having any immediate negative impact that I can think of for people who don't care. And it would loosely couple runtime and build-time environments in a way that people can easily decouple if they want (because in reality, they _are_ coupled in the sense that you usually build in the same place that you test).

I just think back and shudder at all of the support requests we got during the PHP 5.6 -> 7 transition, especially from DrupalVM users, who would install PHP 7 packages on their host machine that would then fail in DrupalVM or cloud environments running PHP 5.6. Encouraging people to restrict PHP versions seems like the best way to avoid that.
